lib/core/CLogger.cc

```diff
@@ -6,61 +6,121 @@
+                                    const args_map& args) {
+        args_map::const_iterator iter{args.find(FORMAT)};
+        if (iter != args.end()) {
+            return boost::log::expressions::stream
+                   << boost::log::expressions::format_date_time<boost::posix_time::ptime>(
+                          boost::log::expressions::attr<boost::posix_time::ptime>(name),
+                          iter->second);
+        } else {
+            return boost::log::expressions::stream
+                   << boost::log::expressions::attr<boost::posix_time::ptime>(name);
+        }
+    }
+};
+
+const std::string CTimeStampFormatterFactory::FORMAT{"format"};
 }
 
 namespace ml {
 namespace core {
 
 CLogger::CLogger()
-    : m_Logger(0), m_Reconfigured(false), m_ProgramName(CProgName::progName()),
+    : m_Reconfigured(false), m_FileAttributeName("File"),
+      m_LineAttributeName("Line"), m_FunctionAttributeName("Function"),
       m_OrigStderrFd(-1), m_FatalErrorHandler(defaultFatalErrorHandler) {
     CCrashHandler::installCrashHandler();
+    // These formatter factories are not needed for programmatic configuration
+    // of the format, but without them formats defined in settings files don't
+    // work correctly.
+    boost::log::register_simple_formatter_factory<ELevel, char>("Severity");
+    boost::log::register_simple_filter_factory<ELevel, char>("Severity");
+    boost::log::register_formatter_factory(
+        "TimeStamp", boost::make_shared<CTimeStampFormatterFactory>());
+    auto loggingCorePtr{boost::log::core::get()};
+    // By default Boost.Log logs in local time but doesn't remember the timezone.
+    // For the case of logging to the ES JVM over a named pipe we want to send
+    // the timestamp as milliseconds since the epoch, which requires either a
+    // UTC timestamp or knowledge of the timezone.  For logging to a log file or
+    // the console we want to print the timezone.  However, logging to a log
+    // file or the console is only for internal development.  The pragmatic
+    // approach here is to always log in UTC.  (This can be revisited in the
+    // future if having console logs in UTC is ever a problem, but any fix would
+    // be non-trivial.)
+    loggingCorePtr->add_global_attribute(boost::log::aux::default_attribute_names::timestamp(),
+                                         boost::log::attributes::utc_clock());
+    loggingCorePtr->add_global_attribute(
+        boost::log::aux::default_attribute_names::process_id(),
+        boost::log::attributes::current_process_id());
+    loggingCorePtr->add_global_attribute(
+        boost::log::aux::default_attribute_names::thread_id(),
+        boost::log::attributes::current_thread_id());
     this->reset();
 }
 
 CLogger::~CLogger() {
-    log4cxx::LogManager::shutdown();
-    m_Logger = nullptr;
+    auto loggingCorePtr{boost::log::core::get()};
+    loggingCorePtr->flush();
+    loggingCorePtr->remove_all_sinks();
```

Comment:
Feels like we could factor out a cc local `resetSink(shared_ptr< sinks::sink >)` function which if it is supplied `nullptr` just calls `flush` and `remove_all_sinks` on `boost::log::core::get()`.